### DISCUSSION TITLE
Threading in C++

### DISCUSSION THREAD

**Discussion by j-balkovec** [community/community#54909] [Programming Help]
### Body

Hi,

I am working on my HashTable assignment and I am required to read a .csv file with a bunch of covid19 data (300K lines). I implemented this, which works so far, it's just very slow.

```
void CovidDB::add_covid_data(std::string const COVID_FILE) {
    std::ifstream file(COVID_FILE);

    if (!file) {
        std::cout << "\n[File ERROR]\n " << COVID_FILE << std::endl;
        std::exit(EXIT_FAILURE);
    }

    std::string line;
    std::getline(file, line); // skip header line

    std::string latest_date_str = "01/01/00"; // initialize to an old date
    std::tm latest_date = {};
    std::istringstream iss(latest_date_str);
    iss >> std::get_time(&latest_date, "%m/%d/%y");

    while (std::getline(file, line)) {
        std::stringstream ss(line);
        std::string country, date_str, cases_str, deaths_str;
        std::getline(ss, date_str, ',');
        std::getline(ss, country, ',');
        std::getline(ss, cases_str, ',');
        std::getline(ss, deaths_str, ',');

        int cases = std::stoi(cases_str);
        int deaths = std::stoi(deaths_str);

        std::tm entry_date = {};
        std::istringstream iss2(date_str);
        iss2 >> std::get_time(&entry_date, "%m/%d/%y");

        if (mktime(&entry_date) > mktime(&latest_date)) {
            latest_date_str = date_str;
            latest_date = entry_date;
        } 

        DataEntry* entry = new DataEntry();
        entry->set_country(country);
        entry->set_date(latest_date_str);
        entry->set_c_cases(cases);
        entry->set_c_deaths(deaths);

        add(entry);
    }

    file.close();
}
```

I've read online that threading could speed this up, so I tried the following, but it doesn't work.

```
void CovidDB::add_covid_data(std::string const COVID_FILE) {
    std::ifstream file(COVID_FILE);

    if (!file) {
        std::cout << "\n[File ERROR]\n " << COVID_FILE << std::endl;
        std::exit(EXIT_FAILURE);
    }

    std::string line;
    std::getline(file, line); // skip header line

    std::string latest_date_str = "01/01/00"; // initialize to an old date
    std::tm latest_date = {};
    std::istringstream iss(latest_date_str);
    iss >> std::get_time(&latest_date, "%m/%d/%y");

    std::vector<std::thread> threads;

    while (std::getline(file, line)) {
        threads.emplace_back([&]() {
            std::stringstream ss(line);
            std::string country, date_str, cases_str, deaths_str;
            std::getline(ss, date_str, ',');
            std::getline(ss, country, ',');
            std::getline(ss, cases_str, ',');
            std::getline(ss, deaths_str, ',');
            std::cout << cases_str << " " << deaths_str << std::endl;
            int cases = std::stoi(cases_str);
            int deaths = std::stoi(deaths_str);

            std::tm entry_date = {};
            std::istringstream iss2(date_str);
            iss2 >> std::get_time(&entry_date, "%m/%d/%y");

            if (mktime(&entry_date) > mktime(&latest_date)) {
                latest_date_str = date_str;
                latest_date = entry_date;
            }

            DataEntry* entry = new DataEntry();
            entry->set_country(country);
            entry->set_date(latest_date_str);
            entry->set_c_cases(cases);
            entry->set_c_deaths(deaths);

            add(entry);
        });
    }

    for (auto& thread : threads) {
        thread.join();
    }

    file.close();
}
```

Could someone tell me where I went wrong because it's saying my stoi failed in the second one but not in the first one. This is my first time working with threading and I am still trying to get the hang of it.

Thank you

Makefile:
```
CXX = g++
CXXFLAGS = -std=c++11 -pthread -g -Wall -Wextra -Werror -pedantic -Wno-unused-parameter -Wno-return-type -Wno-unused-variable
LDFLAGS = -pthread

all: main

main: CovidDB.o main.o
	$(CXX) $(CXXFLAGS) -o $@ $^

CovidDB.o: CovidDB.cpp CovidDB.h
	$(CXX) $(CXXFLAGS) -c $<

main.o: main.cpp CovidDB.h
	$(CXX) $(CXXFLAGS) -c $<

clean:
	rm -f main *.o
```

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by HammoucheSonia** [comment]
#include <mutex>

void CovidDB::add_covid_data(std::string const COVID_FILE) {
    // ... (rest of the code remains the same)

    std::mutex mutex; // Add a mutex for synchronization

    while (std::getline(file, line)) {
        threads.emplace_back([&]() {
            // ... (rest of the code remains the same)

            // Lock the mutex before updating shared data
            std::lock_guard<std::mutex> lock(mutex);

            if (mktime(&entry_date) > mktime(&latest_date)) {
                latest_date_str = date_str;
                latest_date = entry_date;
            }

            // Unlock the mutex after updating shared data
        });
    }

    // ... (rest of the code remains the same)
}

**Comment by RamtinKosari** [comment]
Hi dear @j-balkovec 
The issue you're facing with the second implementation using threads is likely due to the fact that the lambda function you're passing to the thread captures the line variable by reference (&), but the line variable goes out of scope before the lambda function executes.

To fix this, you can capture the line variable by value (=) instead of by reference. Here's the modified code:
```c++
while (std::getline(file, line)) {
    threads.emplace_back([=]() {  // Capture line by value
        // Rest of your code...
    });
}
```
By capturing line by value, each lambda function will have its own copy of the line variable, ensuring that it remains valid even after the loop iteration ends.

Additionally, make sure that the line variable is not used outside the lambda function after the loop, as it will only contain the last line read from the file.

With these changes, the threading should work correctly without any issues. However, keep in mind that using threading in this scenario may not necessarily improve performance significantly, as the bottleneck might be elsewhere (e.g., disk I/O). You can profile your code to identify the actual performance bottlenecks and optimize accordingly.
